sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala

```diff
@@ -82,8 +82,11 @@ case class AQEShuffleReadExec private(
         // `RoundRobinPartitioning` but we don't need to retain the number of partitions.
         case r: RoundRobinPartitioning =>
           r.copy(numPartitions = partitionSpecs.length)
-        case other => throw new IllegalStateException(
-          "Unexpected partitioning for coalesced shuffle read: " + other)
+        case _ =>
+          // Spark plugins may have custom partitioning and may replace this operator
+          // during the postStageOptimization phase, so return UnknownPartitioning here
+          // rather than throw an exception
+          UnknownPartitioning(partitionSpecs.length)
```

Comment:
Hmm, this was asked and answered by @cloud-fan in a comment https://github.com/apache/spark/pull/33541/files#r678751988. So the plugin uses a custom shuffle partitioning?

Reply:
We have a GPU version of `HashPartitioning`. When the above code runs, we have already executed a shuffle on the GPU that has `GpuHashPartitioning` and we need to replace `AQEShufflerReader` with a GPU version.

I am going to contribute some tests that have a Spark plugin that uses custom shuffles so that we can better catch this kind of regressions in the future.  I may have to do this as a separate PR but I will see how much work is involved.

Reply:
Thank you, @andygrove .

Reply:
What if `child.outputPartitioning` is some other pre-defined partitioning such as `SinglePartitioning`? seems we should still throw exception?

Reply:
Another option here would be to introduce some new traits such as `HashPartitioningLike`, as we did with `ShuffleExchangeLike` and `BroadcastExchangeLike` so that plugins can use custom partitioning implementations that Spark can still work with. I will put up an alternate PR using that approach in the next day or two.

Reply:
> What if child.outputPartitioning is some other pre-defined partitioning such as SinglePartitioning?

We can list these unexpected partitionings explicitly here and fail.

Reply:
ya that'll do

Reply:
I added handlikng for `SinglePartition`. I didn't add handling for other `Partitioning` classes such as `BroadcastPartitioning` and `DataSourcePartitionin` because those didn't seem possible in this context but let me know if I should add them.